GSoC 2015: Projected Augmented Reality

micuat opened the pull request:
Hi,

This is a pull request for a GSoC project, Projected Augmented Reality.

Todo:
- ✖ more efficient segmentation/clustering algorithm  
  ➔ **haven't been done yet since this turned out to be complicated; currently inefficient but at least works**
- ✖ make test (stubbed)  
  ➔ **no tests since contributions are mostly sample codes**
- ✖ projector-camera (Kinect) calibration
  - <del>what is the status of the structured light project?</del>  
    ➔ **merged from Roberta's branch, but the stub version**
- ✓ a depth-map capture program
- ✓ uv texture coordinate mapping
- ✓ upload the Unity renderer app
- ✖ Processing/openFrameworks renderer for projection mapping w/o Unity
  ➔ **haven't been done since now it's hugely relying on Unity features**
- ✖ integrate with `surface_matching`  
  ➔ **will be a post-GSoC project**
- update tutorials

RobertaRavanelli committed:
creazione readme

RobertaRavanelli committed:
First API stub

RobertaRavanelli committed:
Second API stubs

RobertaRavanelli committed:
Fixed unused parameters warnings

RobertaRavanelli committed:
Fixed conversion warnings on Windows

RobertaRavanelli committed:
Updated README

micuat committed:
Rgbd image clustering for projection mapping

RobertaRavanelli committed:
Fixed unused parameters warnings and conversion warnings on Windows

RobertaRavanelli committed:
Implemented shadowMasks computation

RobertaRavanelli committed:
Preliminary decoding implementation

micuat committed:
Draft of clustering and projection mapping tutorial

micuat committed:
Doxygen of rgbd clustering

micuat committed:
Format trailing whitespace for rgbd clustering

micuat committed:
Fix include files of clustering example

micuat committed:
Casting and unix compatible functions to pass build

micuat committed:
Replace c++11 functions in rgbd clustering tutorial

micuat committed:
Move texture coordinate mapping from calculatePoints to unwrapTexCoord

RobertaRavanelli committed:
Removed whitespaces and EOF line

micuat committed:
Add CV_EXPORTS; make placeholder for unused parameters for rgbd clustering

micuat committed:
Add rgbd clustering test (stub)

micuat committed:
RgbdCluster to inherit RgbdFrame (no impl so far); move meshing functions
to RgbdClusterMesh

micuat committed:
Delete member variables duplicate of inhereted ones

micuat commented:
[PR 275](https://github.com/FelixMDenis/opencv_contrib/blob/master/modules/surface_matching/src/ppf_helpers.cpp) has a .ply exporter. Implementing .ply and .obj exporter in separate modules is not good, I think

micuat committed:
Cleanup inheritance of RgbdCluster(Mesh)

micuat committed:
RgbdCluster with Ptr<RgbdFrame> to save memory for common matrices

bmagyar commented:
@micuat Indeed. The student preferred to have it this way last year and there wasn't more time to move it outside. Those should be a library call, or we should make it a new module "meshtools" ?

micuat commented:
Thanks @bmagyar, I'm even thinking of merging `surface_matching` to `rgbd` since they have similar features (`rgbd` has ICP for odometry).

Currently I'm trying to refactor `rgbd` so that `depthTo3d` and other functions can be applied to `RgbdFrame` directly, instead of passing several matrices as function arguments. I'm trying to keep backward compatibility but probably need other peoples' ideas as well (in `rgbd_projection_dev` branch, as I don't want to mess up this PR before midterm evaluation).
https://github.com/micuat/opencv_contrib/blob/rgbd_projection_dev/modules/rgbd/include/opencv2/rgbd.hpp

RobertaRavanelli committed:
Completed decoding implementation and code cleaning

RobertaRavanelli committed:
Formatted code

RobertaRavanelli committed:
Removed EOF line from README

bmagyar commented:
Yes! I did some exploratory work before thinking about how to make the API of surface matching more convenient and I found the need to use depthTo3d to move from depth images to a set of 3d points. If you have time for it, feel fee to merge it. I may do it myself by the end of summer if all goes well.

micuat committed:
Add opennl dependency

micuat committed:
Refactor rgbd; add roi to cluster; OpenNL texture unwrapping (not
working); ply output

micuat committed:
Use plane coefficients for unwrapping if plane

micuat committed:
Merge branch 'MidTermPullRequest' into rgbd_projection_dev

micuat committed:
rgbd + structured_light for sensor-projector calibration sample (halfway)

micuat committed:
Update sensor-projector calibrator; seems to be a problem with gray code decoding

micuat committed:
Fix sensor-projector calibrator; now seems working except y-axis is
flipped??

micuat committed:
XML output compatible with RoomAlive; still tweaking flipping issue

micuat committed:
Change visibility in structured_light module

micuat committed:
Update projection mapping tutorial

micuat committed:
Change float to double for calib3d compatibility; add debug output

micuat committed:
Sensor-projector calibration changed to a semi-auto method by Moreno and
Taubin

micuat committed:
Save camera intrinsics; more general way to find homography

micuat committed:
Fix for OpenNI2 calibration

micuat committed:
More accurate sensor-projector calibration by homography fitting`

micuat committed:
Compute local homography to find better projector lens distortion

micuat committed:
More stability for sensor-projector calibration by preventing exception at
findHomography and stereoCalibrate

micuat committed:
Point cloud reconstruction example using structured light

micuat committed:
Render depth map in color

micuat committed:
Rectify depth image to projector space and load to clustering; problem is
projector calibration's focal length result is bad and plane is distorted,
which cannot be extracted by rgbd plane

micuat committed:
Separate structured light correspondence acquisition example and
calibration example

micuat committed:
Use Subdiv2D Delaunay triangulation for meshing and OpenNL for texture
mapping

micuat committed:
Sensor_projector_calibration saves encoded correspondence maps and cluster_projection loads them for mesh
warping

micuat committed:
Delaunay & uv unwrapping in projector space

micuat committed:
Modification in rgbd samples regarding user interface

micuat committed:
Output warped image in projector space

micuat committed:
Move sensor calibration parameters to an external yml file

micuat committed:
Switch for OpenNI or not

micuat committed:
Mesh generation for projector-(color)camera

micuat committed:
Improved plane extraction from segmentation results

micuat committed:
Eliminate depth/structured light discontinuity

micuat committed:
Normalize uv map to 0-1

micuat committed:
Change search window

micuat committed:
Refactoring meshing classes to RgbdMesh

micuat committed:
Grabber function replaces duplicate lines in calibration sample

micuat committed:
Output error when uv unwrapping called but not build with OpenNL

micuat committed:
Renaming and dropping non maintained samples

micuat committed:
Add command line argument parser to rgbd examples (so that Unity can set
parameters)

micuat committed:
Abort meshing sample when non openni2 camera is used

micuat committed:
Annotation and cleanup of meshing codes

micuat committed:
Change variable name to avoid warnings

micuat committed:
Add Unity tutorial project

micuat committed:
Delete trailing whitespaces

micuat committed:
Fix warning/error in rgbd samples

micuat committed:
Automatically make structured-light window fullscreen on projector display

micuat committed:
Rearrange RgbdMesh initialization order

micuat committed:
Drop mesh save option to output texcoord mesh

micuat committed:
Static cast for VideoCapture argument to avoid Linux gcc error

micuat committed:
Add rgbd projection mapping tutorials; need update

micuat committed:
Update first 3 pages of projection Unity tutorial

micuat committed:
Update projection mapping instructions

micuat committed:
More instructions on quad projection mapping
